In [4]:
import json
#import konlpy
import pandas as pd
import numpy as np

#import elice_challenge as ec

In [5]:
train_data = json.load(open('../elice/data_r3/train_data.json'))
test_data = json.load(open('../elice/data_r3/test.input.json'))
grading_data = json.load(open('../elice/data_r3/grading.input.json'))

In [6]:
print(len(train_data))
print(len(test_data))
print(len(grading_data))

76500
4500
9000


In [10]:
total_list = []

#int(len(train_data)/10)
for i in range(len(train_data)):
    total_list.append(train_data[i]['datetime'])
    total_list.append(train_data[i]['text'])
    total_list.append(train_data[i]['title'])
    total_list.append(train_data[i]['writing'])

In [12]:
total_list = np.reshape(total_list, (len(train_data), 4))
total_list

array([['20161212',
        '[앵커]청와대 민정수석을 지냈던 우병우씨의 소재를 놓고 설왕설래가 한창입니다. 지난주 청문회 때 출석요구서 수령을 거부해 나오지 않았지요. 국회는 오는 19일 5차 청문회에 우씨를 다시 부르기로 했지만, 그의 소재가 파악되지 않는 한, 불가능한 일입니다. 일각에선 그에게 현상금까지 거는 상황까지 벌어지고 있는데요. 오늘(12일) 국회 발제에선 이 문제를 놓고 얘기해보도록 하겠습니다.   ]네, 그렇습니다. 지난주 \'최순실 국정농단 청문회\'에 우병우씨는 모습을 드러내지 않았습니다. 다른 사람들처럼 출석요구서를 전달받고 불출석사유를 낸 것도 아니고, 아예 잠적을 해버려서, 요구서 전달조차 안 된 상황이었던 겁니다. 어떻게 보면 질이 가장 안 좋은 상황이라고 볼 수 있는 거죠.보다 못한 몇몇분들이 "우병우를 잡겠다"고 하면서 현상금을 걸었습니다.정봉주 전 의원, 안민석 의원이 각각 500만원, 김성태 \'최순실 국정조사 특위위원장\'이 100만원, 또 주진우 가 200만원, 이렇게 도합 1300만원이 걸린 겁니다.심지어는 많은 시민들이 \'나도 동참하고 싶다\'면서 계좌를 열어라! 하셨다는군요. 그래서 일종의 \'펀딩\' 형식으로 액수가 계속 불어날 것 같다고 합니다.다른 분들은 그런 줄 알겠는데, 안민석 의원이 500만원을 걸었다, 이건 좀 충격적인데요. 저번에 소개해드렸지만, 장시호 잡겠다고 저랑 같이 야밤에 여기저기 돌아다닐 때, 배가 너무 고파서, 같이 24시간 하는 식당 가서 떡만둣국을 먹었거든요. 잘 먹고 나서 계산하려고 하니까, 갑자기 "지갑을 차에 두고 왔다"고 하는 겁니다.이날 밥값을 제가 냈습니다. 아직 영수증도 갖고 있습니다. 그러셨던 분이 500만원을 걸었답니다.자, 아무튼 제가 현상금을 맨 처음 제안한 정봉주 전 의원에게 전화해봤습니다. 뭔가 그럴듯한 제보가 들어왔는지 말이죠. 그랬더니 이렇게 말합니다.[정봉주/전 의원 : 종로구 옥인동 소재에 우병우 닮은 사람이 마당에서 서성인다는 걸 

In [17]:
train_df = pd.DataFrame(total_list, columns=['datetime', 'text', 'title', 'writing'])
train_df.tail()

datetime                                               text  \
76495  20160708  "박선숙·김수민 영장청구한 날 발표, 공정한지 의문…회계 책임자도 조사해야"    ...   
76496  20161010  "새누리당, 대통령 측근 지키기 넘어 청와대 부속실인 듯"【서울=<언론사명>】박대로...   
76497  20160216  김정현 김재두 김경록 출마 채비  = 국민의당 대변인 6명 중 광주·전남 출신 3명...   
76498  20161212  지난 9일 국회에서 대통령 박근혜에 대한 탄핵소추안이 가결돼 헌법재판소의 결정을 기...   
76499  20161207  제3자 뇌물제공죄 적용가능성수뢰액 1억넘으면 10년이상 중형‘최순실 국정농단 사건’...   

                                          title writing  
76495         국민의당 "檢, 새누리 정치자금법 위반 혐의 철저 조사해야"    연합뉴스  
76496                추미애 "새누리, 대통령 측근·친인척 만 지켜"     뉴시스  
76497                  국민의당 대변인 3명 광주·전남서 총선 출마     뉴시스  
76498  [월요광장-권대봉 고려대 교수] 대통령과 참모진에게 필요한 세 가지 교육   헤럴드경제  
76499          미르·K스포츠재단 기금성격 원점서 재검토…대가성 입증 주력   헤럴드경제

In [18]:
train_df2 = train_df.groupby('writing')
train_df2.count()

datetime   text  title
writing                        
JTBC         2416   2416   2416
KBS 뉴스       6285   6285   6285
MBN          7399   7399   7399
TV조선         1359   1359   1359
YTN          4089   4089   4089
경향신문         1535   1535   1535
국민일보         5594   5594   5594
노컷뉴스          471    471    471
뉴시스         12363  12363  12363
동아일보         4981   4981   4981
매일경제         3135   3135   3135
서울신문         4057   4057   4057
아시아경제         559    559    559
연합뉴스         7863   7863   7863
연합뉴스TV       3408   3408   3408
오마이뉴스        1934   1934   1934
채널A          2218   2218   2218
파이낸셜뉴스       4181   4181   4181
헤럴드경제        2653   2653   2653

### 언론사 수
- 언론사 19가지

### 언론사별 데이터 수
- JTBC : 2416
- KBS 뉴스 : 6285
- MBN : 7399
- TV조선 : 1359
- YTN : 4089
- 경향신문 : 1535
- 국민일보 : 5594
- 노컷뉴스 : 471
- 뉴시스 : 12363
- 동아일보 : 4981
- 매일경제 : 3135
- 서울신문 : 4057
- 아시아경제 : 559
- 연합뉴스 : 7863
- 연합뉴스TV : 3408
- 오마이뉴스 : 1934
- 채널A : 2218
- 파이낸셜뉴스 : 4181
- 헤럴드경제 : 2653

In [22]:
import math, sys
from konlpy.tag import Twitter
import jpype

class BayesianFilter:
    """ 베이지안 필터 """
    def __init__(self):
        self.words = set() # 출현한 단어 기록
        self.word_dict = {} # 카테고리마다의 출현 횟수 기록
        self.category_dict = {} # 카테고리 출현 횟수 기록
    # 형태소 분석하기 --- (※1)
    def split(self, text):
        results = []
        twitter = Twitter()
        # 단어의 기본형 사용
        malist = twitter.pos(text, norm=True, stem=True)
        for word in malist:
            # 어미/조사/구두점 등은 대상에서 제외 
            if not word[1] in ["Josa", "Eomi", "Punctuation"]:
                results.append(word[0])
        return results
    # 단어와 카테고리의 출현 횟수 세기 --- (※2)
    def inc_word(self, word, category):
        # 단어를 카테고리에 추가하기
        if not category in self.word_dict:
            self.word_dict[category] = {}
        if not word in self.word_dict[category]:
            self.word_dict[category][word] = 0
        self.word_dict[category][word] += 1
        self.words.add(word)
    def inc_category(self, category):
        # 카테고리 계산하기
        if not category in self.category_dict:
            self.category_dict[category] = 0
        self.category_dict[category] += 1
    
    # 텍스트 학습하기 --- (※3)
    def fit(self, text, category):
        """ 텍스트 학습 """
        word_list = self.split(text)
        for word in word_list:
            self.inc_word(word, category)
        self.inc_category(category)
    
    # 단어 리스트에 점수 매기기--- (※4)
    def score(self, words, category):
        score = math.log(self.category_prob(category))
        for word in words:
            score += math.log(self.word_prob(word, category))
        return score
    
    # 예측하기 --- (※5)
    def predict(self, text):
        best_category = None
        max_score = -sys.maxsize 
        words = self.split(text)
        score_list = []
        for category in self.category_dict.keys():
            score = self.score(words, category)
            score_list.append((category, score))
            if score > max_score:
                max_score = score
                best_category = category
        return best_category, score_list
    # 카테고리 내부의 단어 출현 횟수 구하기
    def get_word_count(self, word, category):
        if word in self.word_dict[category]:
            return self.word_dict[category][word]
        else:
            return 0
    # 카테고리 계산
    def category_prob(self, category):
        sum_categories = sum(self.category_dict.values())
        category_v = self.category_dict[category]
        return category_v / sum_categories
        
    # 카테고리 내부의 단어 출현 비율 계산 --- (※6)
    def word_prob(self, word, category):
        n = self.get_word_count(word, category) + 1 # ---(※6a)
        d = sum(self.word_dict[category].values()) + len(self.words)
        return n / d

In [23]:
bf = BayesianFilter()

for i in range(1, int(len(train_data)/10)):
    bf.fit(train_df[i-1:i]['text'], train_df[i-1:i]['writing'])
    

NameError: name 'jpype' is not defined